design/one-pager-managed-resource-api-design.md

```diff
@@ -0,0 +1,398 @@
+* _Claim_. The Crossplane representation of a request for the allocation of a managed resource. Resource claims
+  typically represent the need for a managed resource that implements a particular protocol. MySQLInstance and
+  RedisCluster are examples of resource claims.
+* _Provider_. Cloud provider such as GCP, AWS, Azure offering IaaS, cloud networking, and managed services.
+* _Spec_. The sub-resource of Kubernetes resources that represents the desired state of the user.
+* _Status_ The sub-resource of Kubernetes resources that represents the most up-to-date information about the resource.
+
+For other terms, see [glossary].
+
+## Background
+Crossplane extends Kubernetes heavily and the maintainers always try to adopt the patterns that are already seen in 
+Kubernetes resources, notably declarative style API. However, it's not always intuitive for everyone to see what goes to
+where in a way that is consistent and just as expected by the consumers of those resources.
+
+This document tries to capture how the bridge between Crossplane and provider is shaped. So, the features of managed
+resources that are generic to all of them are not included here, for example reclaim policies. This is more about the
+things that a developer should keep in mind when they try to implement a new managed resource.
+
+A few questions that this document needs to help answering:
+- What goes into `Spec` and what goes into `Status` of a managed resource?
+- Should _all_ information from the provider be represented in the Custom Resource?
+- What parts of the Custom Resource should be allowed to be consumed by the consumers?
+- Do we represent _desired_ state, _observed_ state or both?
+- How should we differentiate the fields that are used by provider from Crossplane's fields?
+
+While this document is heavily opinionated, each provider can wildly vary between each other. So, treat this document as
+a starting point, try to stick to it for consistency but deviate if required and open an issue about why you had to
+deviate so that we can reassess and improve it.
+
+## Guiding Principles
+
+Our users are the stakeholders on how we make decisions. There are two types of users for managed resources:
+* Input Users: Users who configure/create/manage the resource. Crossplane Claim Controller, Cluster Admin are a few
+  examples of this type. We want our `Spec` to be sole place for them to configure everything about the resource.
+* Output Users: Users who refer to/query/use/extract information from the managed resource. We want them to be able to
+  perform all read-only operations on the given resource as if they have the cloud provider's API available. This can
+  include configurations and metadata information about the resource. We also want them to access only `Status`
+  sub-resource of the managed resource because this is where the most up-to-date information about the resource is
+  stored and we are able to limit RBAC of these users to access only `Status`.
```

Comment:
> We also want them to access only Status sub-resource of the managed resource because this is where the most up-to-date information about the resource is stored and we are able to limit RBAC of these users to access only Status.

Does this need updating? It's true that we can use RBAC to limit users to viewing the status, but we can also use RBAC to limit users to viewing (and not updating) the entire object.

Reply:
Good catch, removed that part.